drivers/i2s/i2s_ll_stm32.c

```diff
@@ -0,0 +1,1170 @@
+	struct i2s_stm32_data *const dev_data = DEV_DATA(dev);
+	size_t mem_block_size;
+	int ret;
+
+	ret = queue_get(&stream->mem_block_queue, &stream->mem_block,
+			&mem_block_size);
+	if (ret < 0) {
+		return ret;
+	}
+	k_sem_give(&stream->sem);
+
+	/* Assure cache coherency before DMA read operation */
+	DCACHE_CLEAN(stream->mem_block, mem_block_size);
+
+	if (stream->master) {
+		LL_I2S_SetTransferMode(cfg->i2s, LL_I2S_MODE_MASTER_TX);
+	} else {
+		LL_I2S_SetTransferMode(cfg->i2s, LL_I2S_MODE_SLAVE_TX);
+	}
+
+	/* remember active TX DMA channel (used in callback) */
+	active_dma_tx_channel[stream->dma_channel] = dev;
+
+	ret = start_dma(dev_data->dev_dma, stream->dma_channel,
+			&stream->dma_cfg,
+			stream->mem_block,
+			(void *)LL_SPI_DMA_GetRegAddr(cfg->i2s),
+			stream->cfg.block_size);
+	if (ret < 0) {
+		SYS_LOG_ERR("Failed to start TX DMA transfer: %d", ret);
+		return ret;
+	}
+
+	LL_I2S_EnableDMAReq_TX(cfg->i2s);
+
+	LL_I2S_EnableIT_ERR(cfg->i2s);
+	LL_I2S_Enable(cfg->i2s);
+
+	return 0;
+}
+
+static void rx_stream_disable(struct stream *stream, struct device *dev)
+{
+	const struct i2s_stm32_cfg *cfg = dev->config->config_info;
+	struct i2s_stm32_data *const dev_data = DEV_DATA(dev);
+	struct device *dev_dma = dev_data->dev_dma;
+
+	LL_I2S_Disable(cfg->i2s);
+	LL_I2S_DisableDMAReq_RX(cfg->i2s);
+	LL_I2S_DisableIT_ERR(cfg->i2s);
+
+	dma_stop(dev_dma, stream->dma_channel);
+	if (stream->mem_block != NULL) {
+		k_mem_slab_free(stream->cfg.mem_slab, &stream->mem_block);
+		stream->mem_block = NULL;
+	}
+
+	active_dma_rx_channel[stream->dma_channel] = NULL;
+}
+
+static void tx_stream_disable(struct stream *stream, struct device *dev)
+{
+	const struct i2s_stm32_cfg *cfg = dev->config->config_info;
+	struct i2s_stm32_data *const dev_data = DEV_DATA(dev);
+	struct device *dev_dma = dev_data->dev_dma;
+
+	LL_I2S_Disable(cfg->i2s);
+	LL_I2S_DisableDMAReq_TX(cfg->i2s);
+	LL_I2S_DisableIT_ERR(cfg->i2s);
+
+	dma_stop(dev_dma, stream->dma_channel);
+	if (stream->mem_block != NULL) {
+		k_mem_slab_free(stream->cfg.mem_slab, &stream->mem_block);
+		stream->mem_block = NULL;
+	}
+
+	active_dma_tx_channel[stream->dma_channel] = NULL;
+}
+
+static void rx_queue_drop(struct stream *stream)
+{
+	size_t size;
+	void *mem_block;
+
+	while (queue_get(&stream->mem_block_queue, &mem_block, &size) == 0) {
+		k_mem_slab_free(stream->cfg.mem_slab, &mem_block);
+	}
+
+	k_sem_reset(&stream->sem);
+}
+
+static void tx_queue_drop(struct stream *stream)
+{
+	size_t size;
+	void *mem_block;
+	unsigned int n = 0;
+
+	while (queue_get(&stream->mem_block_queue, &mem_block, &size) == 0) {
+		k_mem_slab_free(stream->cfg.mem_slab, &mem_block);
+		n++;
+	}
+
+	for (; n > 0; n--) {
+		k_sem_give(&stream->sem);
+	}
+}
+
+#ifdef CONFIG_I2S_1
+static struct device DEVICE_NAME_GET(i2s_stm32_1);
+
+static void i2s_stm32_irq_config_func_1(struct device *dev);
+
+static const struct i2s_stm32_cfg i2s_stm32_config_1 = {
+	.i2s = (SPI_TypeDef *) CONFIG_SPI_1_BASE_ADDRESS,
```

Comment:
How does this work with device tree?
It means that SPI_1 should be enabled ?

Reply:
Well, in that case, why not doing it in this PR? It should not be a major effort.
Besides having several IPs located at same address make it an interesting test for dts scripts.

Reply:
Well, this was just because I'm currently not very familiar with dts stuff in zephyr. Nevertheless,I agree it should not be so huge effort. I probably would need some help from your side. For the moment I'm just pushing a new commit with the other changes. I'll come back on this dts topic soon...

Reply:
@mnkp 
I think that @erwango is correct.
Yesterday I was compiling this driver for Nucleo, with I2S enabled but SPI disabled.
And it does not work because it does not find CONFIG_SPI_1_BASE_ADDRESS.
At the moment it is needed to enable both I2S and SPI to make it working, while I guess
it is not expected. As a user I would like to enable I2S only.

Reply:
The message I was trying to convey is a little bit different. The name of the hardware module according to the documentation is SPI and we probably need only one representation of this module in DTS. There is no really need to define I2S in DTS. At least for now. From the hardware point of view it is the same peripheral, same base address, IRQs, etc. 

And yes, to use I2S driver we would need to ensure that necessary instance of SPI hardware module is enabled in DTS otherwise all the defines generated by DTS will be missing.

Later when we refine DTS concept and have automatic generation of driver instances in driver source code this will need to change. But from what I've noticed no one knows yet how to do it properly, the discussion is ongoing, so it's a little bit too early to try to implement it.

Reply:
@mnkp 

> There is no really need to define I2S in DTS [...] it's a little bit too early to try to implement it.

There is already some part of DTS used in a lot of STM32 drivers, such as UART, SPI, I2C, ...
This is not generated instantiation yet, but at least we are able to gather configuration bits
from dts files and input to drivers with the help of .fixup files.
Request for this driver is to achieve the same status.